e2e: TestKuke_AttachDetach_KeepsTaskRunning exits 1 after clean detach (read/write routines exited)

## Repro
On `origin/main` (commit `0a7b433`):

```bash
sudo /usr/local/go/bin/go test -run TestKuke_AttachDetach_KeepsTaskRunning -timeout 600s ./e2e -count=1
```

Reproduced twice this session — once with #146 applied, then with the branch stashed back to clean `origin/main`. Same failure both times, so it is unrelated to that PR.

## Observed
The socket-staging path now succeeds (so this is a different failure mode from the closed #116), `pkg/attach` connects, and the Ctrl+] Ctrl+] detach filter cleanly tears down — but `kuke attach` exits 1 instead of 0:

```
=== RUN   TestKuke_AttachDetach_KeepsTaskRunning
    e2e_kuke_attach_test.go:206: kuke attach exited with code 1 (want 0)
        output:
        [96mTo detach, press ^] twice[0m
        export PS1="(sbsh-$SBSH_TERM_ID) $PS1"
        ^]^G[93m
        Detached[0m
        Error: close requested: read/write routines exited
--- FAIL: TestKuke_AttachDetach_KeepsTaskRunning (14.0s)
```

Note the `Detached` banner prints (the filter fired), then `pkg/attach.Run` returns the `read/write routines exited` error which `kuke attach` surfaces as exit 1.

## Root cause
Not yet pinpointed. Candidate shortlist for the next picker:

- `cmd/kuke/attach/attach.go` — the layer that wraps `sbsh/pkg/attach.Run`. The clean-detach path likely returns a non-nil sentinel (`io.EOF` or similar) that the wrapper treats as an error rather than a normal exit.
- `sbsh/pkg/attach` itself (vendor / module dep) — its `Run` may emit a non-nil error on the user-initiated detach path that callers are expected to filter.
- Any post-detach cleanup in `cmd/kuke/attach` that races the read/write goroutine teardown.

This is precisely the "verify in-process detach behaves end-to-end" check that #119 was filed to do — #119 is a re-run TODO from #115's review, not a bug report, so this is the first issue describing the actual failure mode now that socket-staging (#116) is unblocked.

## Suggested fix
Have the attach wrapper distinguish a clean user-initiated detach (Ctrl+] Ctrl+]) from an unexpected I/O teardown and return nil on the former. The detach filter already has the signal — we just need the exit path to honor it instead of bubbling the goroutine-shutdown error. A follow-up should fold the verification into #119 once the wrapper's exit semantics are settled.

Cross-refs: #119 (open re-run TODO), #116 (closed, different failure mode).

--- a/cmd/kuke/attach/attach.go
+++ b/cmd/kuke/attach/attach.go
@@ -135,12 +135,36 @@ func runAttach(cmd *cobra.Command, _ []string) error {
 	}
 
 	run := resolveRun(cmd)
-	return run(cmd.Context(), attach.Options{
+	if runErr := run(cmd.Context(), attach.Options{
 		SocketPath: result.HostSocketPath,
 		Stdin:      os.Stdin,
 		Stdout:     os.Stdout,
 		Stderr:     os.Stderr,
-	})
+	}); runErr != nil && !isCleanDetach(runErr) {
+		return runErr
+	}
+	return nil
+}
+
+// isCleanDetach reports whether err returned by the in-process attach
+// loop describes a normal session end rather than an unrecoverable
+// failure. The Ctrl+] Ctrl+] detach filter inside sbsh's pkg/attach
+// triggers a controller-level Detach RPC; the embedded read/write
+// goroutines then exit cleanly, but their teardown is reported as the
+// error "close requested: read/write routines exited" — sbsh does not
+// expose a sentinel for either layer in its public pkg/errors. The
+// remote terminal closing the connection (workload exited, peer hung
+// up) ends up on the same path, which from `kuke attach`'s perspective
+// is also a benign session end. Surface a clean exit code 0 in those
+// cases; any other failure (socket missing, RPC error, context done)
+// arrives without this signature and still bubbles up.
+func isCleanDetach(err error) bool {
+	if err == nil {
+		return false
+	}
+	msg := err.Error()
+	return strings.Contains(msg, "close requested") &&
+		strings.Contains(msg, "read/write routines exited")
 }
 
 // pickAttachableContainer enumerates the cell's containers and returns the

--- a/cmd/kuke/attach/attach_test.go
+++ b/cmd/kuke/attach/attach_test.go
@@ -75,6 +75,18 @@ func (r *runCapture) fn(_ context.Context, opts sbshattach.Options) error {
 	return nil
 }
 
+// runReturning is the mirror of runCapture for tests that want pkg/attach.Run
+// to return a specific error (rather than the clean-detach default).
+type runReturning struct {
+	calls int
+	err   error
+}
+
+func (r *runReturning) fn(_ context.Context, _ sbshattach.Options) error {
+	r.calls++
+	return r.err
+}
+
 func newCmdWithCtx(t *testing.T, fc *fakeClient, run *runCapture) *cobra.Command {
 	t.Helper()
 
@@ -263,6 +275,66 @@ func TestAttach_ExplicitContainer_Attachable_Succeeds(t *testing.T) {
 	}
 	if run.opts.SocketPath != testHostSocket {
 		t.Errorf("Options.SocketPath = %q, want %q", run.opts.SocketPath, testHostSocket)
+	}
+}
+
+// TestAttach_RunReturnsCleanDetachError covers the regression from #147:
+// when the user presses Ctrl+] Ctrl+], sbsh's controller fires the
+// Detach RPC and the embedded read/write copy goroutines exit, but that
+// teardown surfaces as "close requested: read/write routines exited"
+// from pkg/attach.Run. The wrapper must recognise that signature as a
+// benign session end and return nil so `kuke attach` exits 0.
+func TestAttach_RunReturnsCleanDetachError(t *testing.T) {
+	t.Cleanup(viper.Reset)
+
+	fc := &fakeClient{
+		attachContainerFn: func(_ v1beta1.ContainerDoc) (kukeonv1.AttachContainerResult, error) {
+			return kukeonv1.AttachContainerResult{HostSocketPath: testHostSocket}, nil
+		},
+	}
+	cleanDetachErr := errors.New("close requested: read/write routines exited")
+	run := &runReturning{err: cleanDetachErr}
+	cmd := newCmdWithCtx(t, fc, nil)
+	cmd.SetContext(context.WithValue(cmd.Context(), attachcmd.MockRunKey{}, attachcmd.RunFn(run.fn)))
+	cmd.SetArgs([]string{
+		"--realm", "r1", "--space", "s1", "--stack", "st1", "--cell", "c1",
+		"--container", "work",
+	})
+
+	if err := cmd.Execute(); err != nil {
+		t.Fatalf("Execute returned %v on clean-detach error, want nil", err)
+	}
+	if run.calls != 1 {
+		t.Fatalf("attach.Run called %d times, want 1", run.calls)
+	}
+}
+
+// TestAttach_RunReturnsRealError ensures that any pkg/attach.Run error
+// not matching the clean-detach signature still bubbles up — the wrapper
+// must not swallow real failures like socket-not-found or RPC errors.
+func TestAttach_RunReturnsRealError(t *testing.T) {
+	t.Cleanup(viper.Reset)
+
+	fc := &fakeClient{
+		attachContainerFn: func(_ v1beta1.ContainerDoc) (kukeonv1.AttachContainerResult, error) {
+			return kukeonv1.AttachContainerResult{HostSocketPath: testHostSocket}, nil
+		},
+	}
+	bootErr := errors.New("wait on ready: dial unix: connection refused")
+	run := &runReturning{err: bootErr}
+	cmd := newCmdWithCtx(t, fc, nil)
+	cmd.SetContext(context.WithValue(cmd.Context(), attachcmd.MockRunKey{}, attachcmd.RunFn(run.fn)))
+	cmd.SetArgs([]string{
+		"--realm", "r1", "--space", "s1", "--stack", "st1", "--cell", "c1",
+		"--container", "work",
+	})
+
+	err := cmd.Execute()
+	if err == nil {
+		t.Fatalf("Execute returned nil on real error, want %v", bootErr)
+	}
+	if !errors.Is(err, bootErr) {
+		t.Fatalf("error %v does not unwrap to bootErr", err)
 	}
 }
 

